src/eventstream-proxy/eventstream-proxy.base.ts

```diff
@@ -91,7 +91,17 @@ export abstract class EventStreamProxyBase extends WebSocketEventsBase {
       this.url,
       this.topic,
       events => {
-        this.queueTask(() => this.processEvents(events));
+        this.queueTask(async () => {
+          while (true) {
+            try {
+              const result = await this.processEvents(events);
+              return result;
+            } catch (err) {
+              this.logger.error(`Error processing event (retrying in 3s): ${err}`);
+              await new Promise(resolve => setTimeout(resolve, 3000));
```

Comment:
Are there any errors here that are unrecoverable? If so, how do we handle that?

Reply:
Yes, the most prominent one I can think of is if you deploy an ERC721 that does not support `tokenUri` (although FWIW the README calls out that `tokenUri` must be implemented). The first time you perform a transfer, it will hang forever.

There could be others that I'm not thinking of.

Reply:
I noted a few details about this above as well: https://github.com/hyperledger/firefly-tokens-erc20-erc721/pull/122#issuecomment-1433563545

Reply:
ok - so dead letter handling is hard...

The right thing to do is to have categorization of "bad data" failures (log + ack and move on) separate from "infrastructure issue" failures (retry indefinitely).

If you just have plain errors that could be anything, and they're all in one bucket, the best I can think of is:

Given that the only two resolutions we have are:
1. Consume the message
2. Retry consumption

The most sophisticated code based solution (which might be best put into the FireFly SDK?) is:
- Keep track in-memory of the ID of each failure, on the basis we expect it to be re-delivered
- Keep a counter on the number of times that failure recurs
- Have a threshold after which that failure is considered acceptable and move on

However, the danger here is in the (pretty likely) case that the messages can't be processed because you have a dependency that's broken - such as being able to connect to FF Core right now - you end up in a situation where you do want to try indefinitely.

Reply:
So... FireFly is not really involved here. This is the token connector listening to blockchain events from evmconnect. But the analysis is probably still about the same.

The only "infrastructure issue" that I'm particularly trying to address is retrieving the token URI by querying the contract after each transfer event. So I could try to target the retry to that case more specifically. It's also possible that retrieving the token URI on every transfer should not be the responsibility of the token event loop (but that would imply a larger change at this point).

Reply:
I'm removing this from scope and will open a separate PR for this piece only.